Assets/Scripts/BoardManager.cs

```diff
@@ -22,47 +44,168 @@ public class BoardManager : MonoBehaviour
+                validPosition = new List<string>() { "top", "right", "bottom" };
+            }
+            else if (validPosition[randomIndex] == "bottom")
+            {
+                roomPosition.Add("bottom");
+                y--;
+                m.Add(new Vector2(x, y));
+                validPosition = new List<string>() { "right", "bottom", "left" };
+            }
+            else if (validPosition[randomIndex] == "left")
+            {
+                roomPosition.Add("left");
+                x--;
+                m.Add(new Vector2(x, y));
+                validPosition = new List<string>() { "top", "bottom", "left" };
+            }
+        }
+
+        //Generate first room part
+        GenerateRoomPart(x, y, m, 0);
 
-        for (int x = 1; x < columns - 1; x++)
+        //Generate other room parts
+        int index = 1;
+        foreach (var pos in roomPosition)
         {
-            for (int y = 1; y < rows - 1; y++)
+            if (pos == "top")
+            {
+                y += rowsCount;
+                GenerateRoomPart(x, y, m, index); //top
+            }
+            else if (pos == "right")
+            {
+                x += columnsCount;
+                GenerateRoomPart(x, y, m, index); //right
+            }
+            else if (pos == "bottom")
+            {
+                y -= rowsCount;
+                GenerateRoomPart(x, y, m, index); //bottom
+            }
+            else if (pos == "left")
             {
-                gridPositions.Add(new Vector3(x, y, 0f));
+                x -= columnsCount;
+                GenerateRoomPart(x, y, m, index); //left
             }
+            index++;
         }
     }
 
-    void BoardSetup()
+    void GenerateRoomPart(int xNumber, int yNumber, List<Vector2> m, int roomPos)
     {
-        boardHolder = new GameObject("Board").transform;
+        List<string> wallPositions = CheckWhereToInstantiateWall(m, roomPos);
 
-        for (int x = -1; x < columns + 1; x++)
+        for (int x = xNumber; x < columnsCount + xNumber + 1; x++)
         {
-            for (int y = -1; y < rows + 1; y++)
+            for (int y = yNumber; y < rowsCount + yNumber + 1; y++)
             {
                 GameObject toInstantiate = floorTiles[Random.Range(0, floorTiles.Length)];
-                if (x == -1 || x == columns || y == -1 || y == rows)
+
+                Vector3 vector3 = new Vector3(x, y, 0f);
+
+                if ((x == xNumber && wallPositions.Contains("left")) || (x == (columnsCount + xNumber) && wallPositions.Contains("right")) || (y == yNumber && wallPositions.Contains("bottom")) || (y == (rowsCount + yNumber) && wallPositions.Contains("top")))
                 {
-                    toInstantiate = outerWallTiles[Random.Range(0, outerWallTiles.Length)];
+                    //Don't add a wall if there is already one
+                    if (!outerWallGridPositions.Contains(vector3) && !gridPositions.Contains(vector3))
```

Comment:
I don't know how this condition can be true if `CheckWhereToInstanciateWall` is correct.
If it isn't you should instead fix `CheckWhereToInstanciateWall`